File: nova_lang/lib/nova/set.ex
defmodule Nova.Set do
  @moduledoc """
  Set operations for Nova runtime.
  Sets are represented as MapSets.
  """

  def empty, do: MapSet.new()

  def singleton(x), do: MapSet.new([x])

  def insert(x, set), do: MapSet.put(set, x)

  def member(x, set), do: MapSet.member?(set, x)

  def delete(x, set), do: MapSet.delete(set, x)

  def union(s1, s2), do: MapSet.union(s1, s2)

  def difference(s1, s2), do: MapSet.difference(s1, s2)

  def intersection(s1, s2), do: MapSet.intersection(s1, s2)

  def from_foldable(list), do: MapSet.new(list)

  def to_unfoldable(set), do: MapSet.to_list(set)

  def map(f, set) do
    set
    |> MapSet.to_list()
    |> Enum.map(f)
    |> MapSet.new()
  end

  def map_maybe(f, set) do
    set
    |> MapSet.to_list()
    |> Enum.flat_map(fn x ->
      case f.(x) do
        {:just, v} -> [v]
        :nothing -> []
      end
    end)
    |> MapSet.new()
  end

  def filter(pred, set) do
    set
    |> MapSet.to_list()
    |> Enum.filter(pred)
    |> MapSet.new()
  end

  def size(set), do: MapSet.size(set)

  def is_empty(set), do: MapSet.size(set) == 0

  def find_min(set) do
    list = MapSet.to_list(set)
    case list do
      [] -> :nothing
      _ -> {:just, Enum.min(list)}
    end
  end
end

File: runtime/set.erl
-module(set).
-export([delete/2, difference/2, empty/0, insert/2, intersection/2,
         member/2, singleton/1, size/1, union/2, fromFoldable/1]).

%% Using Erlang's sets module internally

%% Empty set
empty() -> sets:new().

%% Insert element
insert(X, Set) -> sets:add_element(X, Set).

%% Delete element
delete(X, Set) -> sets:del_element(X, Set).

%% Check membership
member(X, Set) -> sets:is_element(X, Set).

%% Set size
size(Set) -> sets:size(Set).

%% Create singleton set
singleton(X) -> sets:from_list([X]).

%% Set union
union(S1, S2) -> sets:union(S1, S2).

%% Set intersection
intersection(S1, S2) -> sets:intersection(S1, S2).

%% Set difference
difference(S1, S2) -> sets:subtract(S1, S2).

%% Create set from list
fromFoldable(List) -> sets:from_list(List).
